src/modules/ekf2/EKF/aid_sources/optical_flow/optical_flow_control.cpp

```diff
@@ -38,45 +38,69 @@
 
 #include "ekf.h"
 
+#include <ekf_derivation/generated/compute_flow_xy_innov_var_and_hx.h>
+
 void Ekf::controlOpticalFlowFusion(const imuSample &imu_delayed)
 {
-	if (_flow_buffer) {
-		_flow_data_ready = _flow_buffer->pop_first_older_than(imu_delayed.time_us, &_flow_sample_delayed);
+	if (!_flow_buffer || (_params.flow_ctrl != 1)) {
+		stopFlowFusion();
+		return;
 	}
 
+	VectorState H;
+
 	// New optical flow data is available and is ready to be fused when the midpoint of the sample falls behind the fusion time horizon
-	if (_flow_data_ready) {
+	if (_flow_buffer->pop_first_older_than(imu_delayed.time_us, &_flow_sample_delayed)) {
+		flowSample &flow_sample = _flow_sample_delayed;
+
 		const int32_t min_quality = _control_status.flags.in_air
 					    ? _params.flow_qual_min
 					    : _params.flow_qual_min_gnd;
 
-		const bool is_quality_good = (_flow_sample_delayed.quality >= min_quality);
-		const bool is_magnitude_good = _flow_sample_delayed.flow_rate.isAllFinite()
-					       && !_flow_sample_delayed.flow_rate.longerThan(_flow_max_rate);
+		const bool is_quality_good = (flow_sample.quality >= min_quality);
+		const bool is_magnitude_good = flow_sample.flow_rate.isAllFinite()
+					       && !flow_sample.flow_rate.longerThan(_flow_max_rate);
 
 		bool is_tilt_good = true;
 
 #if defined(CONFIG_EKF2_RANGE_FINDER)
 		is_tilt_good = (_R_to_earth(2, 2) > _params.range_cos_max_tilt);
-
 #endif // CONFIG_EKF2_RANGE_FINDER
 
-		if (is_quality_good
-		    && is_magnitude_good
-		    && is_tilt_good) {
-			// compensate for body motion to give a LOS rate
-			calcOptFlowBodyRateComp(imu_delayed);
-			_flow_rate_compensated = _flow_sample_delayed.flow_rate - _flow_sample_delayed.gyro_rate.xy();
-
-		} else {
-			// don't use this flow data and wait for the next data to arrive
-			_flow_data_ready = false;
-			_flow_rate_compensated.setZero();
+		if (is_quality_good && is_magnitude_good) {
+			calcOptFlowBodyRateComp(imu_delayed, flow_sample);
```

Comment:
I think we shouldn't really care about the flow magnitude and quality to compute the angular rate compensation.

Reply:
Likely not, but I wasn't sure if we can depend on synchronized gyro always being populated in all cases.

Reply:
fixed